packages/compiler/src/core/types.ts

```diff
@@ -308,13 +400,23 @@ export interface Scalar extends BaseType, DecoratedType, TemplatedTypeBase {
    */
   derivedScalars: Scalar[];
 
+  constructors: Map<string, ScalarConstructor>;
   /**
-   * Late-bound symbol of this model type.
+   * Late-bound symbol of this scalar type.
    * @internal
    */
   symbol?: Sym;
 }
 
+// TODO: should we just call that Constructor or NamedConstructor for future proof?
+export interface ScalarConstructor extends BaseType {
+  kind: "ScalarConstructor";
+  node: ScalarConstructorNode;
+  name: string;
+  scalar: Scalar;
+  parameters: FunctionParameter[];
```

Comment:
Should we seperate `FunctionParameters` that is currently used to describe decorators function parmaters and accept `valueof` vs `Scalar Constructor` already infer that it must take in a value.